### PR TITLE
mpl2: ensure that sa converges if a valid result was generated at any point

### LINKED ISSUE #5996: [ERROR MPL-0040] Failed on cluster frontend/icache
### Describe the bug

Similar to https://github.com/The-OpenROAD-Project/OpenROAD/issues/5884.

A small change to the pin placements in a SRAM causes MPL to fail:

```
[INFO ORD-0030] Using 32 thread(s).
HierRTLMP Flow enabled...
rtl_macro_placer -halo_width 19 -halo_height 19 -report_directory .//objects/asap7/BoomTile/base/rtlmp -target_util 0.20
Floorplan Outline: (0.0, 0.0) (1500.0, 1500.0),  Core Outline: (2.052, 2.16) (1497.96, 1497.96)
	Number of std cell instances: 1819087
	Area of std cell instances: 229568.39
	Number of macros: 62
	Area of macros: 466562.44
	Halo width: 19.00
	Halo height: 19.00
	Area of macros with halos: 950132.19
	Area of std cell instances + Area of macros: 696130.81
	Core area: 2237579.00
	Design Utilization: 0.31
	Core Utilization: 0.13
	Manufacturing Grid: 1

[ERROR MPL-0040] Failed on cluster frontend/icache
Error: macro_place_util.tcl, 83 MPL-0040
Command exited with non-zero status 1
Elapsed time: 11:18.15[h:]min:sec. CPU time: user 8840.85 sys 119.27 (1321%). Peak memory: 9239248KB.
```

I tried to use deltaDebug.py to reduce the test case, but wasn't successful after 24h.

### Expected Behavior

No error

### Environment

```shell
OpenROAD v3.0-1620-gf61d7a1e
```

### To Reproduce

untar https://drive.google.com/file/d/1YOaN5KWzO6whiU7uvhL5lzA-7DB_DjgT/view?usp=sharing

```
run-me-BoomTile-asap7-base.sh
```

### Relevant log output

_No response_

### Screenshots

_No response_

### Additional Context

_No response_

### LINKED ISSUE #6031: Macro placement fails with SETUP_HOLD_MARGIN=0
### Describe the bug

With SETUP_HOLD_MARGIN=-1300, macro placement works, with SETUP_HOLD_MARGIN=0, as in this example, macro placement fails.

untar https://drive.google.com/file/d/1ns113IUJjeH4gIxOk6r0ARbu6XBkvYbb/view?usp=sharing

```
OpenROAD v2.0-16787-gcd519bb5e
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
[INFO ORD-0030] Using 48 thread(s).
HierRTLMP Flow enabled...
rtl_macro_placer -halo_width 19 -halo_height 19 -report_directory .//objects/asap7/BoomTile/1/rtlmp -target_util 0.24
Floorplan Outline: (0.0, 0.0) (2000.0, 2000.0),  Core Outline: (2.052, 2.16) (1998.0, 1998.0)
        Number of std cell instances: 2012393
        Area of std cell instances: 241154.34
        Number of macros: 72
        Area of macros: 814224.38
        Halo width: 19.00
        Halo height: 19.00
        Area of macros with halos: 1461710.00
        Area of std cell instances + Area of macros: 1055378.75
        Core area: 3983592.75
        Design Utilization: 0.26
        Core Utilization: 0.08
        Manufacturing Grid: 1

[ERROR MPL-0040] Failed on cluster frontend/bpd/banked_predictors_0/btb
Error: macro_place_util.tcl, 83 MPL-0040
Command exited with non-zero status 1
Elapsed time: 17:22.26[h:]min:sec. CPU time: user 9617.43 sys 95.13 (931%). Peak memory: 23057736KB.
```

### Expected Behavior

Macro placement should work

### Environment

```shell
OpenROAD v2.0-16787-gcd519bb5e
```

### To Reproduce

See above

### Relevant log output

_No response_

### Screenshots

_No response_

### Additional Context

_No response_

### LINKED ISSUE #6055: macro placement fails
### Describe the bug

with SYNTH_HIERARCHICAL=0, number of instances is half of that of SYNTH_HIERARCHICAL=1, so it was worth trying to turn it off, but macro placement fails. https://github.com/The-OpenROAD-Project/megaboom/pull/188

untar https://drive.google.com/file/d/1wvt-ThCrWKvSeo4jKmxmgpnbLtjV_XRE/view?usp=sharing

```
$ ./run-me-BoomTile-asap7-1.sh 
OpenROAD v2.0-16664-g9ba039016 
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
rtl_macro_placer -halo_width 19 -halo_height 19 -report_directory .//objects/asap7/BoomTile/1/rtlmp -target_util 0.24
Floorplan Outline: (0.0, 0.0) (2000.0, 2000.0),  Core Outline: (2.052, 2.16) (1998.0, 1998.0)
	Number of std cell instances: 1046614
	Area of std cell instances: 127052.54
	Number of macros: 72
	Area of macros: 810855.88
	Halo width: 19.00
	Halo height: 19.00
	Area of macros with halos: 1394272.12
	Area of std cell instances + Area of macros: 937908.44
	Core area: 3983592.75
	Design Utilization: 0.24
	Core Utilization: 0.04
	Manufacturing Grid: 1


[ERROR MPL-0040] Failed on cluster (root)_glue_logic_1_0_0
Error: macro_place_util.tcl, 76 MPL-0040
Command exited with non-zero status 1
Elapsed time: 39:50.24[h:]min:sec. CPU time: user 6409.13 sys 687.58 (296%). Peak memory: 16733712KB.
```

### Expected Behavior

Successful macro placement

### Environment

```shell
OpenROAD v2.0-16664-g9ba039016
```

### To Reproduce

See above.

### Relevant log output

_No response_

### Screenshots

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by AcKoucher**
Resolves #5996,
Resolves #6031,
Resolves #6055.

When I implemented #5908 I missed one fundamental thing: the cluster placement result (and all the other SoftMacro annealing results) is not just the sequence pair itself but rather a combination of the sequence pair along with the specific shapes chosen during a perturbation iteration. _I.e., the same sequence pair can represent different floorplan arrangements depending on the clusters' shapes._

As this only applies for SoftMacro annealing, even though the logic for storing the best valid result happens inside the base class SimulatedAnnealingCore, we can only actually use that result in the derived SACoreSoftMacro, because we need SoftMacro APIs to set the shapes that we cached. Because of that, I thought it was time to also remove the methods for floorplan centralization (which are cluster placement specific) from the base class as well.

The SoftMacro annealing result now consists of both a sequence pair and the shapes of that floorplan arrangement.

### CODE REVIEW

**Comment by AcKoucher** [comment]
Running secure-ci.

**Comment by rovinski** [comment]
Great PR message 👍
Really helps with clarity and understanding.

**Comment by AcKoucher** [comment]
I'm glad it's helpful :)

File: src/mpl2/src/SimulatedAnnealingCore.h [hunk 1]
```diff
@@ -126,15 +120,21 @@ class SimulatedAnnealingCore
 
   // Initialize the SA worker
   virtual void initialize() = 0;
-  // Run FastSA algorithm
+  virtual void run() = 0;
```
Comment (maliberty) [inline src/mpl2/src/SimulatedAnnealingCore.h:123]: Is fastSA the public access and run() for internal dispatch, or is there some case for clients to call run directly?  It is public.

Comment (AcKoucher) [reply src/mpl2/src/SimulatedAnnealingCore.h:123]: Actually it's the opposite. fastSA() should be private as it's called inside run(). I'll fix it

**Comment by AcKoucher** [comment]
Secure-CI is fine.

File: src/mpl2/src/SACoreSoftMacro.cpp [hunk 2]
```diff
@@ -101,6 +101,29 @@ SACoreSoftMacro::SACoreSoftMacro(
   logger_ = logger;
 }
 
+void SACoreSoftMacro::run()
+{
+  if (graphics_) {
+    graphics_->startSA();
+  }
+
+  best_valid_result_ = std::make_unique<SoftResult>();
```
[the remark this answers is not in the document]
Comment (AcKoucher) [reply src/mpl2/src/SACoreSoftMacro.cpp:110]: The result for hard macro annealing is a different combination of things:

SoftResult = sequence pair + shapes
HardResult = sequence pair + orientation (not implemented yet)

That said, the HardResult is out of the scope of this PR with regards to what it is trying to fix (which is a problem only in cluster placement).

**Review by maliberty — APPROVED** [review]

**Pull request merged**